src/overlays/actors/ovl_Elf_Msg2/z_elf_msg2.c

```diff
@@ -21,23 +50,177 @@ const ActorInit Elf_Msg2_InitVars = {
+};
+
+void ElfMsg2_SetupAction(ElfMsg2* this, ElfMsg2ActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+/**
+ * Checks a scene flag - if flag is set, the actor is killed and function returns 1. Otherwise returns 0.
+ * Can also set a switch flag from params while killing.
+ */
+s32 ElfMsg2_KillCheck(ElfMsg2* this, GlobalContext* globalCtx) {
+
+    if ((this->actor.posRot.rot.y > 0) && (this->actor.posRot.rot.y < 0x41) &&
+        (Flags_GetSwitch(globalCtx, this->actor.posRot.rot.y - 1))) {
+        // "Mutual destruction"
+        LOG_STRING("共倒れ", "../z_elf_msg2.c", 171);
+        if ((this->actor.params >> 8 & 0x3F) != 0x3F) {
+            Flags_SetSwitch(globalCtx, (this->actor.params >> 8 & 0x3F));
+        }
+        Actor_Kill(&this->actor);
+        return 1;
+    }
+
+    else if ((this->actor.posRot.rot.y == -1) && (Flags_GetClear(globalCtx, this->actor.room))) {
+        // "Mutual destruction 2"
+        LOG_STRING("共倒れ２", "../z_elf_msg2.c", 182);
+        if (((this->actor.params >> 8) & 0x3F) != 0x3F) {
+            Flags_SetSwitch(globalCtx, ((this->actor.params >> 8) & 0x3F));
+        }
+        Actor_Kill(&this->actor);
+        return 1;
+    } else if (((this->actor.params >> 8) & 0x3F) == 0x3F) {
+        return 0;
+    }
+
+    else if (Flags_GetSwitch(globalCtx, ((this->actor.params >> 8) & 0x3F))) {
+        // "Mutual destruction"
+        LOG_STRING("共倒れ", "../z_elf_msg2.c", 192);
+        Actor_Kill(&this->actor);
+        return 1;
+    }
+    return 0;
+}
+
+void ElfMsg2_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ElfMsg2* this = THIS;
+
+    osSyncPrintf(VT_FGCOL(CYAN) " Elf_Msg2_Actor_ct %04x\n\n" VT_RST, this->actor.params);
+    if (!ElfMsg2_KillCheck(this, globalCtx)) {
+        if ((this->actor.posRot.rot.x > 0) && (this->actor.posRot.rot.x < 8)) {
+            this->actor.unk_1F = this->actor.posRot.rot.x - 1;
+        }
+        Actor_ProcessInitChain(thisx, sInitChain);
+        if (this->actor.posRot.rot.y >= 0x41) {
+            ElfMsg2_SetupAction(this, ElfMsg2_Idle);
+        } else {
+            ElfMsg2_SetupAction(this, ElfMsg2_WaitForRead);
+            this->actor.flags |= 0x00040001; // Make actor targetable and Navi-checkable
+            this->actor.textId = ElfMsg2_GetMessageId(this);
+        }
+        this->actor.shape.rot.x = this->actor.shape.rot.y = this->actor.shape.rot.z = 0;
+    }
+}
+
+void ElfMsg2_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Init.s")
+s32 ElfMsg2_GetMessageId(ElfMsg2* this) {
+    return (this->actor.params & 0xFF) + 0x100;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Elf_Msg2/ElfMsg2_Destroy.s")
+/**
+ * Runs while Navi text is up. Kills the actor upon closing the text box unless rot.z == 1, can also set a switch flag
+ * from params.
+ */
+void ElfMsg2_Read(ElfMsg2* this, GlobalContext* globalCtx) {
```

Comment:
"Read" is probably not a good action name for this actor, since it's not really what the actor itself is doing